tools/loader/ModelCompiler.cpp

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Loader.h"
+
+#include "glow/Converter/TypeAToTypeBFunctionConverter.h"
+#include "glow/Graph/Nodes.h"
+#include "glow/Importer/Caffe2ModelLoader.h"
+#include "glow/Importer/ONNXModelLoader.h"
+
+#include "llvm/Support/CommandLine.h"
+
+using namespace glow;
+
+namespace {
+
+/// Model compiler options.
+llvm::cl::OptionCategory modelCompilerCat("Model Compiler Options");
+
+llvm::cl::list<std::string> modelInputs(
+    "model-input", llvm::cl::ZeroOrMore,
+    llvm::cl::desc(
+        "For ONNX models the inputs of the graph can be inferred   \n"
+        "automatically and hence this option is not mandatory.     \n"
+        "For Caffe2 models the graph definition does not contain   \n"
+        "the description of the inputs and hence must be provided  \n"
+        "explicitly using this option. One or more model inputs    \n"
+        "are provided using the following format:                  \n"
+        "   -model-input=<inputName1>,<inputType1>,<inputShape1>   \n"
+        "   -model-input=<inputName2>,<inputType2>,<inputShape2>   \n"
+        "   ....................................................   \n"
+        "For quantized types the format is slightly different since\n"
+        "the scale and offset parameters should also be provided:  \n"
+        "   -model-input=<name>,<type>,<scale>,<offset>,<shape>    \n"
+        "For example we can can provide one or more inputs:        \n"
+        "   -model-input=input_03_data,float,[1]                   \n"
+        "   -model-input=data_bias,int32,[1,32,32]                 \n"
+        "   -model-input=data,int8q,0.123,-13,[1,10]               \n"
+        "The supported types are:                                  \n"
+        "   - float, float16 (floating point types)                \n"
+        "   - int32, int64 (integer types)                         \n"
+        "   - int8q, int16q, int32q (integer quantized types)      \n"
+        "   - bool (logic type)\n"),
+    llvm::cl::value_desc("name,type,shape"), llvm::cl::cat(modelCompilerCat));
+
+} // namespace
+
+/// Parse the strings and get the model input names and types.
+static void getModelInputs(std::vector<std::string> &inputNames,
```

Comment:
I feel like this function for parsing input names and types could be useful as a re-usable utility function, so that new similar tools can be built on top of it. Please consider moving it somewhere where it could be re-used. Maybe somewhere in `Caffe2ModelLoader.cpp` (or `Loader.cpp`) as these flags seems to be relevant only to loaders for the time being?

Reply:
I`ll do it if you tell me explicitly where to put it :).
The function is currently written such that is packaged together with the CLI option.
My 2 cents: I wouldn`t put in Caffe2ModelLoader because it could be necessary later for ONNX models (don`t know). If I put in Loader.cpp then the image classifier will have both "-model-input-name" and "-model-input" (kind of confusing).

Reply:
Actually, for consistency I`m thinking about using this utility also in image-classifier. Then I`ll put it in Loader.cpp. The single thing I`m reluctant about is that I`ll replace the "-model-input-name" with "-model-input" and I`ll have to make changes in several places. What do you say?

Reply:
> If I put in Loader.cpp then the image classifier will have both "-model-input-name" and "-model-input" (kind of confusing).

Yeah, but I don't think it is too bad. I'd say let's put it into Loader.cpp.

Reply:
> I'll replace the "-model-input-name" with "-model-input" and I`ll have to make changes in several places. What do you say?

Overall, I kind of like this idea.

Can we also support syntax for `model-input`, where only name (without type and shape) is provided, so that e.g. we could declare `-model-input-name` as an alias for `-model-input` and use the old syntax... Then you don't need to update the existing uses of `model-input-name`. We may also check depending on the loader being used if the specified input has a type shape specified in the command-line or not.

Reply:
I moved the utility function in the Loader class.
I also added the format with default type like this:
-model-input=<name>    with default type 'float' and default shape '[1]'

Now in principle this could be used to alias -model-input-name but I would keep the modifications to image-classifier as a separate PR because I envision there another modification:
- currently the input shape is derived from the input png images: I don`t like this because a user could provide images with the wrong size
- the input shape should be derived from the model itself for ONNX and from CLI for Caffe (just as for model-compiler) and then the image sizes should be checked against the model input size